### PR TITLE
[SPARK-39551][SQL] Add AQE invalid plan check

### PR DESCRIPTION
**Pull request by maryannxue**
### What changes were proposed in this pull request?

This PR adds a check for invalid plans in AQE replanning process. The check will throw exceptions when it detects an invalid plan, causing AQE to void the current replanning result and keep using the latest valid plan.

### Why are the changes needed?

AQE logical optimization rules can lead to invalid physical plans and cause runtime exceptions as certain physical plan nodes are not compatible with others. E.g., `BroadcastExchangeExec` can only work as a direct child of broadcast join nodes, but it could appear under other incompatible physical plan nodes because of empty relation propagation.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added UT.

### CODE REVIEW

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ValidateSparkPlan.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+   * - BroadcastQueryStage only appears as the immediate child and the build side of a broadcast
+   *   hash join or broadcast nested loop join.
+   */
+  private def validate(plan: SparkPlan): Unit = plan match {
+    case b: BroadcastHashJoinExec =>
+      val (buildPlan, probePlan) = b.buildSide match {
+        case BuildLeft => (b.left, b.right)
+        case BuildRight => (b.right, b.left)
+      }
+      if (!buildPlan.isInstanceOf[BroadcastQueryStageExec]) {
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ValidateSparkPlan.scala:48]: if the BroadcastHashJoinExec is from a sort merge join, it's direct child should be BroadcastExchangeExec ? for example, two joins, the first join is BroadcastHashJoinExec and the second join is sort merge join at first then after re-optimze it changes to broadcast join.

Or we can move this rule after the `createQueryStages`

**Comment by cloud-fan** [comment]
Yea this is a framework and we can add more checks to it.

**Comment by cloud-fan** [comment]
Yea this is a framework and we can add more checks to it.

**Comment by dongjoon-hyun** [comment]
I made a backport to branch-3.2 for the comment, https://github.com/apache/spark/pull/37087#issuecomment-1175840358 .

**Comment by dongjoon-hyun** [comment]
This will be a part of Apache Spark 3.2.2.

**Pull request closed without merging**